File: leetcode/0350-intersection-of-two-arrays-ii/solution.py
class Solution:
    def intersect(self, nums1: List[int], nums2: List[int]) -> List[int]:
        if len(nums1) > len(nums2):
            return self.intersect(nums2, nums1)
        
        nums1_counter = dict()
        
        for num1 in nums1:
            nums1_counter[num1] = nums1_counter.get(num1, 0) + 1
        
        intersection = list()
        for num2 in nums2:
            count = nums1_counter.get(num2, 0)
            if count > 0:
                nums1_counter[num2] = count - 1
                intersection.append(num2)
        
        return intersection

File: leetcode/0350-intersection-of-two-arrays-ii/solution.cpp
class Solution {
public:
    vector<int> intersect(vector<int>& nums1, vector<int>& nums2) {
        sort(nums1.begin(), nums1.end());
        sort(nums2.begin(), nums2.end());
        vector<int> ans;
        int n1 = nums1.size();
        int n2 = nums2.size();
        int i1 = 0;
        int i2 = 0;
        while (i1 < n1 && i2 < n2) {
            while (i1 < n1 && i2 < n2 && nums1[i1] == nums2[i2]) {
                ans.push_back(nums1[i1]);
                i1++;
                i2++;
            }
            if (i1 < n1 && i2 < n2 && nums1[i1] < nums2[i2]) {
                while (i1 < n1 && nums1[i1] < nums2[i2])    i1++;
            } else if (i1 < n1 && i2 < n2) {
                while (i2 < n2 && nums1[i1] > nums2[i2])    i2++;
            }
        }
        return ans;
    }
};